mm2src/mm2_main/src/lp_swap/maker_swap_v2.rs

```diff
@@ -0,0 +1,898 @@
+
+// This is needed to have Debug on messages
+#[allow(unused_imports)] use prost::Message;
+
+/// If a certain P2P message is not received, swap will be aborted after this time expires.
+const NEGOTIATION_TIMEOUT_SEC: u64 = 90;
+
+/// Represents events produced by maker swap states.
+#[derive(Debug, PartialEq)]
+pub enum MakerSwapEvent {
+    /// Swap has been successfully initialized.
+    Initialized {
+        maker_coin_start_block: u64,
+        taker_coin_start_block: u64,
+    },
+    /// Started waiting for taker payment.
+    WaitingForTakerPayment {
+        maker_coin_start_block: u64,
+        taker_coin_start_block: u64,
+    },
+    /// Received taker payment info.
+    TakerPaymentReceived {
+        maker_coin_start_block: u64,
+        taker_coin_start_block: u64,
+        taker_payment: TransactionIdentifier,
+    },
+    /// Sent maker payment.
+    MakerPaymentSent {
+        maker_coin_start_block: u64,
+        taker_coin_start_block: u64,
+        maker_payment: TransactionIdentifier,
+    },
+    /// Something went wrong, so maker payment refund is required.
+    MakerPaymentRefundRequired { maker_payment: TransactionIdentifier },
+    /// Taker payment has been confirmed on-chain.
+    TakerPaymentConfirmed {
+        maker_coin_start_block: u64,
+        taker_coin_start_block: u64,
+        maker_payment: TransactionIdentifier,
+        taker_payment: TransactionIdentifier,
+    },
+    /// Maker successfully spent taker's payment.
+    TakerPaymentSpent {
+        maker_coin_start_block: u64,
+        taker_coin_start_block: u64,
+        maker_payment: TransactionIdentifier,
+        taker_payment: TransactionIdentifier,
+        taker_payment_spend: TransactionIdentifier,
+    },
+    /// Swap has been aborted before maker payment was sent.
+    Aborted { reason: String },
+    /// Swap completed successfully.
+    Completed,
+}
+
+/// Represents errors that can be produced by [`MakerSwapStateMachine`] run.
+#[derive(Debug, Display)]
+pub enum MakerSwapStateMachineError {}
+
+/// Dummy storage for maker swap events (used temporary).
+#[derive(Default)]
+pub struct DummyMakerSwapStorage {
+    events: HashMap<Uuid, Vec<MakerSwapEvent>>,
+}
+
+#[async_trait]
+impl StateMachineStorage for DummyMakerSwapStorage {
+    type MachineId = Uuid;
+    type Event = MakerSwapEvent;
+    type Error = MakerSwapStateMachineError;
+
+    async fn store_event(&mut self, id: Self::MachineId, event: Self::Event) -> Result<(), Self::Error> {
+        self.events.entry(id).or_insert_with(Vec::new).push(event);
+        Ok(())
+    }
+
+    async fn get_unfinished(&self) -> Result<Vec<Self::MachineId>, Self::Error> {
+        Ok(self.events.keys().copied().collect())
+    }
+
+    async fn mark_finished(&mut self, _id: Self::MachineId) -> Result<(), Self::Error> { Ok(()) }
+}
+
+/// Represents the state machine for maker's side of the Trading Protocol Upgrade swap (v2).
+pub struct MakerSwapStateMachine<MakerCoin, TakerCoin> {
```

Comment:
With HD wallet, we can have swaps from different addresses running at the same time, should we have information about the address used in here? I think a better approach would be to map swap uuids with the addresses used at the wallet layer to separate concerns but I am just opening this point for discussion in case you have a better idea about this.

Reply:
It seems an interesting task with a variety of possible implementation ways 🙂 I suggest to start discussing it in the call.